mixer/pkg/protobuf/yaml/decoder.go

```diff
@@ -0,0 +1,593 @@
+				mapType.Field[1].GetType() != descriptor.FieldDescriptorProto_TYPE_STRING {
+				dv.err = multierror.Append(dv.err, errors.New("only map<string, string> is supported in expressions"))
+				return
+			}
+
+			// translate map<X, Y> proto3 field type to record Mixer type (map[string]string)
+			if _, ok := dv.out[f.GetName()]; !ok {
+				dv.out[f.GetName()] = make(map[string]string)
+			}
+
+			visitor := &mapVisitor{desc: mapType}
+			for len(v) > 0 {
+				_, _, m := ConsumeField(visitor, v)
+				if m < 0 {
+					dv.err = multierror.Append(dv.err, fmt.Errorf("failed to parse map field %q: %v", f.GetName(), ParseError(m)))
+					return
+				}
+				v = v[m:]
+			}
+
+			dv.out[f.GetName()].(map[string]string)[visitor.key] = visitor.value
+			return
+		}
+		// TODO(kuat): implement sub-message decoding
+		return
+	}
+
+	// fallback into packed repeated encoding
+	if f.IsRepeated() && (f.IsPacked() || f.IsPacked3()) {
+		var m int
+		for len(v) > 0 {
+			switch f.GetType() {
+			case descriptor.FieldDescriptorProto_TYPE_BOOL,
+				descriptor.FieldDescriptorProto_TYPE_INT32,
+				descriptor.FieldDescriptorProto_TYPE_INT64,
+				descriptor.FieldDescriptorProto_TYPE_UINT32,
+				descriptor.FieldDescriptorProto_TYPE_UINT64,
+				descriptor.FieldDescriptorProto_TYPE_SINT32,
+				descriptor.FieldDescriptorProto_TYPE_SINT64,
+				descriptor.FieldDescriptorProto_TYPE_ENUM:
+				var elt uint64
+				elt, m = ConsumeVarint(v)
+				if m < 0 {
+					dv.err = multierror.Append(dv.err, fmt.Errorf("failed to parse packed varint field %q", f.GetName()))
+					return
+				}
+				dv.varint(n, elt)
+
+			case descriptor.FieldDescriptorProto_TYPE_FIXED32,
+				descriptor.FieldDescriptorProto_TYPE_SFIXED32,
+				descriptor.FieldDescriptorProto_TYPE_FLOAT:
+				var elt uint32
+				elt, m = ConsumeFixed32(v)
+				if m < 0 {
+					dv.err = multierror.Append(dv.err, fmt.Errorf("failed to parse packed fixed32 field %q", f.GetName()))
+					return
+				}
+				dv.fixed32(n, elt)
+
+			case descriptor.FieldDescriptorProto_TYPE_FIXED64,
+				descriptor.FieldDescriptorProto_TYPE_SFIXED64,
+				descriptor.FieldDescriptorProto_TYPE_DOUBLE:
+				var elt uint64
+				elt, m = ConsumeFixed64(v)
+				if m < 0 {
+					dv.err = multierror.Append(dv.err, fmt.Errorf("failed to parse packed fixed64 field %q", f.GetName()))
+					return
+				}
+				dv.fixed64(n, elt)
+
+			default:
+				dv.err = multierror.Append(dv.err, fmt.Errorf("unexpected field type %q for packed repeated bytes encoding", f.GetType()))
+				return
+			}
+			v = v[m:]
+		}
+		return
+	}
+
+	dv.err = multierror.Append(dv.err, fmt.Errorf("unexpected field type %q for bytes encoding", f.GetType()))
+}
+
+// wireVisitor is used to process field values on-line
+type wireVisitor interface {
+	varint(Number, uint64)
+	fixed32(Number, uint32)
+	fixed64(Number, uint64)
+	bytes(Number, []byte)
+}
+
+// Below is the code borrowed from golang/protobuf api-v2 branch, intended to be aligned closely once
```

Comment:
Suggest move the code  borrowed from v2 protobuf to a separate sub package. Also add a GitHub link to that source.
When we v2 proto is ga, it will just be an import change.

Reply:
I think api-v2 branch is temporary work, since it will eventually be merged into master.
There are problems with fully isolating the borrowed code into a package:
- it does not have enough extension points (visitor) for us to use as-is
- `Resolver` needs to be modified to conform to the upstream interfaces, which is out of scope of this PR
- it's many months from being released and will likely churn a lot before that.